saliency/xrai.py

```diff
@@ -0,0 +1,555 @@
+        attr: Source attributions for XRAI. XRAI attributions will be same size
+              as the input attr.
+        segs: Input segments as a list of boolean masks. XRAI uses these to
+              compute attribution sums.
+        gain_fun: The function that computes XRAI area attribution from source
+                  attributions. Defaults to _gain_density, which calculates the
+                  density of attributions in a mask.
+        area_perc_th: The saliency map is computed to cover area_perc_th of
+                      the image. Lower values will run faster, but produce
+                      uncomputed areas in the image that will be filled to
+                      satisfy completeness. Defaults to 1.0.
+        verbose: 1 prints information about every added mask. >2 prints
+                 information about skipped masks due to min_pixel_diff.
+                 Defaults to 0, which is silent.
+        min_pixel_diff: Do not consider masks that have difference less than
+                        this number compared to the current mask. Set it to 1
+                        to remove masks that completely overlap with the
+                        current mask.
+        integer_segments: See XRAIParameters. Defaults to True.
+
+    Returns:
+        tuple: saliency heatmap and list of masks or an integer image with
+               area ranks depending on the parameter integer_segments.
+    """
+    output_attr = -np.inf * np.ones(shape=attr.shape, dtype=np.float)
+
+    n_masks = len(segs)
+    current_attr_sum = 0.0
+    current_area_perc = 0.0
+    current_mask = np.zeros(attr.shape, dtype=bool)
+
+    masks_trace = []
+    remaining_masks = {ind: mask for ind, mask in enumerate(segs)}
+
+    added_masks_cnt = 1
+    # While the mask area is less than area_th and remaining_masks is not empty
+    while current_area_perc <= area_perc_th:
+      best_gain = -np.inf
+      best_key = None
+      remove_key_queue = []
+      for mask_key in remaining_masks:
+        mask = remaining_masks[mask_key]
+        # If mask does not add more than min_pixel_diff to current mask, remove
+        mask_pixel_diff = _get_diff_cnt(mask, current_mask)
+        if mask_pixel_diff < min_pixel_diff:
+          remove_key_queue.append(mask_key)
+          if verbose > 2:
+            logging.info("Skipping mask with pixel difference: {:.3g},".format(
+                mask_pixel_diff))
+          continue
+        gain = gain_fun(mask, attr, mask2=current_mask)
+        if gain > best_gain:
+          best_gain = gain
+          best_key = mask_key
+      for key in remove_key_queue:
+        del remaining_masks[key]
+      if len(remaining_masks) == 0:
+        break
+      added_mask = remaining_masks[best_key]
+      mask_diff = _get_diff_mask(added_mask, current_mask)
+      masks_trace.append((mask_diff, best_gain))
+
+      current_mask = np.logical_or(current_mask, added_mask)
+      current_attr_sum = np.sum(attr[current_mask])
```

Comment:
'current_attr_sum' is only used for logging and should be moved inside the 'if' statement.

Reply:
done.